frontend/catalyst/jax_tracer.py

```diff
@@ -828,107 +845,138 @@ def trace_quantum_function(
+            t.transform.__name__ for t in device_program
+        ]
 
-        # (2) - Quantum tracing
-        transformed_results = []
+        tapes, post_processing = apply_transform(
+            qnode_program,
+            device_program,
+            device_modify_measurements,
+            quantum_tape,
+            return_values_flat,
+        )
 
-        with EvaluationContext.frame_tracing_context(ctx, trace):
-            # Set up same device and quantum register for all tapes in the program.
-            # We just need to ensure the qubits are reset in between each.
-            qdevice_p.bind(
-                rtd_lib=device.backend_lib,
-                rtd_name=device.backend_name,
-                rtd_kwargs=str(device.backend_kwargs),
-            )
-            qreg_in = qalloc_p.bind(len(device.wires))
-
-            qnode_transformed = len(qnode_program) > 0
-            for i, tape in enumerate(tapes):
-                # If the program is batched, that means that it was transformed.
-                # If it was transformed, that means that the program might have
-                # changed the output. See `split_non_commuting`
-                if qnode_transformed or device_modify_measurements:
-                    # TODO: In the future support arbitrary output from the user function.
-                    output = tape.measurements
-                    _, trees = jax.tree_util.tree_flatten(output, is_leaf=is_leaf)
-                else:
-                    output = return_values_flat
-                    trees = return_values_tree
-
-                qrp_out = trace_quantum_tape(tape, device, qreg_in, ctx, trace)
-                meas, meas_trees = trace_quantum_measurements(device, qrp_out, output, trees, tape)
-                qreg_out = qrp_out.actualize()
-
-                meas_tracers = [trace.full_raise(m) for m in meas]
-                meas_results = tree_unflatten(meas_trees, meas_tracers)
-
-                # TODO: Allow the user to return whatever types they specify.
-                if qnode_transformed or device_modify_measurements:
-                    assert isinstance(meas_results, list)
-                    if len(meas_results) == 1:
-                        transformed_results.append(meas_results[0])
-                    else:
-                        transformed_results.append(tuple(meas_results))
-                else:
-                    transformed_results.append(meas_results)
 
-                # Reset the qubits and update the register value for the next tape.
-                if len(tapes) > 1 and i < len(tapes) - 1:
-                    for w in device.wires:
-                        qreg_out = reset_qubit(qreg_out, w)
-                    qreg_in = qreg_out
+    # (1.5) If user wanted the `dynamic_one_shot` transformation we apply it by re-tracing the
+    # function using the wrapper loop.
+    if qnode.needs_dynamic_one_shot and device.shots.total_shots != 1:
```

Comment:
`dynamic_one_shot` may also be applied to a single shot tape, but we can neglect that case at first I would say.